packages/axelar-std/src/events.rs

```diff
@@ -1,60 +1,47 @@
 use core::fmt::Debug;
 
-use soroban_sdk::{Env, IntoVal, Topics, Val, Vec};
+use soroban_sdk::{Env, Val, Vec};
 #[cfg(any(test, feature = "testutils"))]
 pub use testutils::*;
 
 pub trait Event: Debug + PartialEq + Sized {
-    fn topics(&self, env: &Env) -> impl Topics + Debug;
+    fn emit(self, env: &Env);
 
-    /// A default empty tuple/vector is used for event data, since majority of events only use topics.
-    fn data(&self, env: &Env) -> impl IntoVal<Env, Val> + Debug {
-        Vec::<Val>::new(env)
-    }
+    fn from_event(env: &Env, topics: Vec<Val>, data: Val) -> Self;
 
-    fn emit(self, env: &Env) {
-        env.events().publish(self.topics(env), self.data(env));
-    }
+    fn schema(env: &Env) -> &'static str;
 }
 
 #[cfg(any(test, feature = "testutils"))]
 mod testutils {
     use soroban_sdk::testutils::Events;
-    use soroban_sdk::{Address, Env, Val, Vec};
+    use soroban_sdk::Env;
 
     use crate::events::Event;
 
-    pub trait EventTestutils: Event {
-        fn matches(self, env: &Env, event: &(Address, Vec<Val>, Val)) -> bool;
-
-        fn standardized_fmt(
-            env: &Env,
-            event: &(soroban_sdk::Address, soroban_sdk::Vec<Val>, Val),
-        ) -> std::string::String;
-    }
-
     pub fn fmt_last_emitted_event<E>(env: &Env) -> std::string::String
     where
-        E: EventTestutils,
+        E: Event,
     {
-        let event = env.events().all().last().expect("no event found");
-        E::standardized_fmt(env, &event)
+        fmt_emitted_event_at_idx::<E>(env, -1)
     }
 
     pub fn fmt_emitted_event_at_idx<E>(env: &Env, mut idx: i32) -> std::string::String
     where
-        E: EventTestutils,
+        E: Event,
     {
         if idx < 0 {
             idx += env.events().all().len() as i32;
         }
 
-        let event = env
+        let (contract_id, topics, data) = env
             .events()
             .all()
             .get(idx as u32)
             .expect("no event found at the given index");
-        E::standardized_fmt(env, &event)
+
+        let event = E::from_event(env, topics, data);
+        std::format!("{:#?}\n\n{:?}\n\n{}", event, contract_id, E::schema(env))
```

Comment:
I specifically avoided parsing the emitted event back into an event. The goal of the golden files is to make sure the emitted data doesn't change, so we don't break consumers downstream. If we take that data and print the parsed event again, we only test that `from_event` works as expected, but it's quite feasible that both serialization and deserialization change, so printing the typed event stays the same, but the emitted event data looks different